Optim foreach cleanup for Adadelta

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* #70229
* #69982
* #69981
* __->__ #69980
* #71335
* #71334
* #71333

Add foreach flag to Adadelta optimizer + cleanup

- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Modified 'test/test_optim:test_multi_tensor_optimizers'

Differential Revision: [D33413059](https://our.internmc.facebook.com/intern/diff/D33413059)

mikaylagawarecki committed:
Optim foreach cleanup for Adadelta

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


Created new file `torch/optim/_multi_tensor_functional.py` which contains foreach functions from `torch/optim/_multi_tensor/_multi_tensor_functional.py`.

Created new file `torch/optim/_single_tensor_functional.py` which contains single tensor functions from `torch/optim/_functional.py`.

All the functions in `torch/optim/_functional.py` now call functions from either file depending on whether foreach=True

Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`





[ghstack-poisoned]

mikaylagawarecki commented on test/test_optim.py:
```diff
@@ -609,6 +691,31 @@ def test_adadelta(self):
                 optimizer(None, lr=1e-2, rho=1.1)
 
+    # new test that test_adadelta can be switched to when merge is complete and multitensor is deleted
+    @skipIfRocm
```
Would it be better to test this flag using the same format that the maximize flag was tested in #67847 and add a dummy foreach flag to the optim/_multitensor classes?

albanD commented on torch/optim/_multi_tensor_functional.py:
```diff
@@ -0,0 +1,37 @@
+r"""Functional interface"""
```
I am actually curious if this is the best way to split these.
Do you think that having these functional implementations live in adadelta.py would be better? (and `_functional.py` imports it from there as exposes them)?
I might make it easier to find a particular implementation no?

mikaylagawarecki replied:
I agree that this would make it easier to find the corresponding functions, but it seems that it might cause a circular import. adadelta.py imports _functional.py to call into the functional api in the Adadelta class, if we store the single tensor and multi tensor functionals in adadelta.py we must import .adadelta in _functional.py. 

Alternatively we could call the corresponding functions directly in the adadelta class and not use the functional api in all the optimizer classes, does that make sense?

albanD replied:
> Alternatively we could call the corresponding functions directly in the adadelta class and not use the functional api in all the optimizer classes, does that make sense?

That sounds good.
I think we want to keep a single functional file to keep an API similar to the one in nn.functional.
But the class version can just call into the local function and the functional simply `from .adadelta import adadelta`.

albanD commented on torch/optim/_functional.py:
```diff
@@ -206,22 +209,25 @@ def adadelta(params: List[Tensor],
-    for (param, grad, square_avg, acc_delta) in zip(params, grads, square_avgs, acc_deltas):
-        if weight_decay != 0:
-            grad = grad.add(param, alpha=weight_decay)
+    if foreach and not torch.jit.is_scripting():
```
```suggestion
    if foreach and not torch.jit.is_scripting():
        multi.multi_tensor_adadelta(params,
                                    grads,
                                    square_avgs,
                                    acc_deltas,
                                    lr=lr,
                                    weight_decay=weight_decay,
                                    rho=rho,
                                    eps=eps)
    else:
        single.single_tensor_adadelta(params,
                                      grads,
                                      square_avgs,
                                      acc_deltas,
                                      lr=lr,
                                      weight_decay=weight_decay,
                                      rho=rho,
                                      eps=eps)
```
```suggestion
    if foreach and not torch.jit.is_scripting():
        func = multi.multi_tensor_adadelta
    else:
        func = single.single_tensor_adadelta
    func(params,
            grads,
            square_avgs,
            acc_deltas,
            lr=lr,
            weight_decay=weight_decay,
            rho=rho,
            eps=eps)
```

albanD commented on torch/optim/_multi_tensor_functional.py:
```diff
@@ -0,0 +1,37 @@
+from torch import Tensor
+from typing import List
+
+def multi_tensor_adadelta(params: List[Tensor],
```
this can be called just `adadelta`

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadeltapy` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Created new file `torch/optim/_multi_tensor_functional.py` which contains foreach functions from `torch/optim/_multi_tensor/_multi_tensor_functional.py`.
- Created new file `torch/optim/_single_tensor_functional.py` which contains single tensor functions from `torch/optim/_functional.py`.
- All the functions in `torch/optim/_functional.py` now call functions from either file depending on whether foreach=True
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False] 





[ghstack-poisoned]

albanD commented on test/test_optim.py:
```diff
@@ -477,6 +477,88 @@ def test_multi_tensor_optimizers(self):
+        if not torch.cuda.is_available():
+            return
+
+        optimizer_pairs_with_flags = [
```
This is very nice to have. Maybe we want an OptimizerInfo at some point @jbschlosser haha

mikaylagawarecki replied:
This function actually just copies `test_multi_tensor_optimizers` and does the same tests when foreach=True, False instead, should I make the below changes to that function as well?

albanD commented on test/test_optim.py:
```diff
@@ -477,6 +477,88 @@ def test_multi_tensor_optimizers(self):
+            # ((optim.Adagrad), dict(weight_decay=1)),
+        ]
+
+        kIterations = 11
```
why 11? 
If we want a random number I would put 3 to reduce time.

albanD commented on test/test_optim.py:
```diff
@@ -477,6 +477,88 @@ def test_multi_tensor_optimizers(self):
+        for opt, params in optimizer_pairs_with_flags:
+            res = []
+            for foreach in (True, False):
+                weight = torch.tensor([[-0.2109, -0.4976], [-0.1413, -0.3420], [-0.2524, 0.6976]],
```
Why hardcode values here instead of torch.rand?
If you want the same values in both loops, you can create these outside.

albanD commented on test/test_optim.py:
```diff
@@ -477,6 +477,88 @@ def test_multi_tensor_optimizers(self):
+                                            torch.nn.Sigmoid(),
+                                            torch.nn.Linear(3, 1),
+                                            torch.nn.Sigmoid())
+                model.to(torch.float64).to(device)
```
nit: `model.to(dtype=torch.float64, device=device)`

albanD commented on test/test_optim.py:
```diff
@@ -477,6 +477,88 @@ def test_multi_tensor_optimizers(self):
+                                            torch.nn.Sigmoid())
+                model.to(torch.float64).to(device)
+
+                pretrained_dict = model.state_dict()
```
Not sure what's the point of this? The model already contains some randomly initialized weights/bias?

mikaylagawarecki replied:
i'm not sure why this was done in the initial test_multitensor function either haha, I changed this bit to use torch.manual_seed instead! :)

albanD commented on test/test_optim.py:
```diff
@@ -477,6 +477,88 @@ def test_multi_tensor_optimizers(self):
+                pretrained_dict['2.bias'] = bias2
+                model.load_state_dict(pretrained_dict)
+
+                params['foreach'] = foreach
```
Aren't you modifying the dictionary contained in optimizer_pairs_with_flags here?
Maybe it doesn't matter as you never re-use it but it looks a bit dangerous no?

albanD commented on test/test_optim.py:
```diff
@@ -477,6 +477,88 @@ def test_multi_tensor_optimizers(self):
+                    loss = output.sum()
+                    loss.backward()
+
+                    if iter == 0:
```
iter doesn't exist here?

albanD commented on test/test_optim.py:
```diff
@@ -609,6 +691,31 @@ def test_adadelta(self):
 
+    # new test that test_adadelta can be switched to when merge is complete and multitensor is deleted
+    @skipIfRocm
+    def test_adadelta_new(self):
```
What is this test covering that is not covered above?

mikaylagawarecki replied:
It is the same test, the above one tests optim.Adadelta and optim_mt.Adadelta, this one tests foreach=True and False instead (if we go ahead with deleting the _multitensor directory, I was intending to add a pr to the top of the stack that copies the logic from `test_*_new` into `test_*`, does that make sense?

albanD commented on torch/optim/_functional.py:
```diff
@@ -196,6 +198,7 @@ def adadelta(params: List[Tensor],
              grads: List[Tensor],
              square_avgs: List[Tensor],
              acc_deltas: List[Tensor],
+             foreach: bool = False,
```
nit: why isn't this one kwarg only with the others below?
Sorry if I already asked and forgot the answer.

mikaylagawarecki replied:
ah this was because `torch/distributed` calls into this functional API and uses torchscript with it. kwarg only arguments that have a default aren't supported by torchscript so I added `foreach` before the `*` since we want the default

albanD commented on torch/optim/adadelta.py:
```diff
@@ -102,13 +108,97 @@ def step(self, closure=None):
+             grads: List[Tensor],
+             square_avgs: List[Tensor],
+             acc_deltas: List[Tensor],
+             foreach: bool = False,
```
nit: kwarg only

albanD commented on torch/optim/adadelta.py:
```diff
@@ -102,13 +108,97 @@ def step(self, closure=None):
+         weight_decay=weight_decay)
+
+
+def single_tensor_adadelta(params: List[Tensor],
```
This and the one below are internal I guess? (needs to start with _)

albanD commented on torch/optim/_functional.py:
```diff
@@ -206,22 +209,16 @@ def adadelta(params: List[Tensor],
-    for (param, grad, square_avg, acc_delta) in zip(params, grads, square_avgs, acc_deltas):
-        if weight_decay != 0:
-            grad = grad.add(param, alpha=weight_decay)
+    adadelta_fn(params,
```
I think the doc here can be moved to the one in adadelta.py and the only thing left in this file will be `from .adadelta import adadelta`

albanD reviewed (commented):
Looks quite good.
It is nice to have both implementations next to each other!

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadeltapy` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Created new file `torch/optim/_multi_tensor_functional.py` which contains foreach functions from `torch/optim/_multi_tensor/_multi_tensor_functional.py`.
- Created new file `torch/optim/_single_tensor_functional.py` which contains single tensor functions from `torch/optim/_functional.py`.
- All the functions in `torch/optim/_functional.py` now call functions from either file depending on whether foreach=True
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False] 





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False] 





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False] 





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False] 





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False] 





[ghstack-poisoned]

mikaylagawarecki commented on test/test_optim.py:
```diff
@@ -432,43 +432,31 @@ def test_multi_tensor_optimizers(self):
         ]
 
-        kIterations = 11
+        kIterations = 4
```
bumping to 4 instead of 3 since 1 iteration is the no-op

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False] 





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False] 





[ghstack-poisoned]

mikaylagawarecki commented on torch/optim/adadelta.py:
```diff
@@ -102,13 +108,111 @@ def step(self, closure=None):
+    if foreach and torch.jit.is_scripting(): 
+        raise RuntimeError('torch.jit.script not supported with foreach optimizers')
+
+    if foreach and not torch.jit.is_scripting(): 
```
had to leave the not torch.jit.is_scripting() here because I tried many variations of this check and they all raised torchscript error that func is undefined when func is set in the else branch (specific error was `Runtime error: undefined value func`)

variations I tried:
```if foreach and torch.jit.is_scripting(): 
        raise RuntimeError('torch.jit.script not supported with foreach optimizers')

    if foreach: 
        func = _multi_tensor_adadelta
    else:
        func = _single_tensor_adadelta
```
and 

```
    if foreach:
        if torch.jit.is_scripting():
            raise RuntimeError('torch.jit.script not supported with foreach optimizers')
        else: # with and without else statement
            func = _multi_tensor_adadelta
    else:
        func = _single_tensor_adadelta
```

and 

```
    if foreach and not torch.jit.is_scripting():
        if torch.jit.is_scripting():
            raise RuntimeError('torch.jit.script not supported with foreach optimizers')
        else: # with and without else statement
            func = _multi_tensor_adadelta
    else:
        func = _single_tensor_adadelta
```

albanD replied:
note that the `not` is not at the right place in the last one ;)

In any case, the current version is good enough I think

albanD commented on torch/distributed/optim/functional_adadelta.py:
```diff
@@ -22,6 +22,7 @@ def __init__(
         rho: float = 0.9,
         eps: float = 1e-6,
         weight_decay: float = 0.0,
+        foreach: bool = False,
```
cc @mrshenli is this ok? Do you want this flexibility on the distributed side or are you always passing a single parameter and want to keep this as False?

albanD reviewed (approved):
Looks good!

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False] 





[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33413059).

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False]

Differential Revision: [D33413059](https://our.internmc.facebook.com/intern/diff/D33413059)

[ghstack-poisoned]

albanD commented on torch/optim/adadelta.py:
```diff
@@ -79,7 +81,11 @@ def step(self, closure=None):
+                                                   group['rho'],
+                                                   group['eps'],
+                                                   group['weight_decay'],
+                                                   group['foreach'])
```
Sorry I missed that in previous reviews...
We want to make sure that checkpoint saved with an older version of pytorch still work here.
I think the simplest way to do that is to add to this class:
```python
def __setstate__(self, state):
    super().__setstate__(state)
    for group in self.param_groups:
        if 'forearch' not in group:
            group['foreach'] = False
```

If you want to test this. You can create an instance of this optimizer. Get the state dict from it. Delete the foeach entries in it and then load_state_dict.
And you should still be able to step with that optimizer after loading this state dict.

mikaylagawarecki replied:
Fixed!

albanD reviewed (commented):
Looks good! Only the serialization issue and this is good to go!

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Added a test `test_optimizers_foreach_flag` which replicates `test_multi_tensor_optimizers` in `test/test_optim.py`
- Add a test `test_adadelta_new` that replicates the behavior of `test_adadelta` but with `foreach` flag instead of using the multitensor adadleta class. If we delete `_multitensor/` we could replace `test_adadelta` with this

Remaining TODO:

- [ ] single_tensor adadelta supports complex but multitensor does not, need to integrate the singletensor logic in multitensor and switch the `test_adadelta_complex` to test for foreach in [True, False]

Differential Revision: [D33413059](https://our.internmc.facebook.com/intern/diff/D33413059)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Modified 'test/test_optim:test_multi_tensor_optimizers'

Differential Revision: [D33413059](https://our.internmc.facebook.com/intern/diff/D33413059)

[ghstack-poisoned]

albanD commented on torch/optim/adadelta.py:
```diff
@@ -58,9 +60,14 @@ def __init__(self, params, lr=1.0, rho=0.9, eps=1e-6, weight_decay=0):
         super(Adadelta, self).__init__(params, defaults)
 
+    def __setstate__(self, state):
+        super(Adadelta, self).__setstate__(state)
```
nit: super() should be enough here.

albanD reviewed (approved):
Thanks for the update!

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Modified 'test/test_optim:test_multi_tensor_optimizers'

Differential Revision: [D33413059](https://our.internmc.facebook.com/intern/diff/D33413059)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Modified 'test/test_optim:test_multi_tensor_optimizers'

Differential Revision: [D33413059](https://our.internmc.facebook.com/intern/diff/D33413059)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33413059).

mikaylagawarecki committed:
Update on "Optim foreach cleanup for Adadelta"


- Merged `torch/optim/adadelta.py` and `torch/optim/_multitensor/adadelta.py` into `torch/optim/adadelta.py`
- Moved adadelta functional forms from `torch/optim/_functional.py` and `torch/optim/_multi_tensor/_functional.py` to `torch/optim/adadelta.py`
- `torch/optim/_functional.py` just imports from `torch/optim/adadelta.py`
- Modified 'test/test_optim:test_multi_tensor_optimizers'

Differential Revision: [D33413059](https://our.internmc.facebook.com/intern/diff/D33413059)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33413059).
